src/Mod/PartDesign/App/FeatureRefine.cpp

```diff
@@ -0,0 +1,139 @@
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+#include <Standard_Failure.hxx>
+#endif
+
+#include <App/Application.h>
+#include <App/FeaturePythonPyImp.h>
+#include <Base/Parameter.h>
+#include <Mod/Part/App/modelRefine.h>
+
+#include "FeatureRefine.h"
+#include "FeaturePy.h"
+
+
+using namespace PartDesign;
+
+namespace PartDesign
+{
+PROPERTY_SOURCE(PartDesign::FeatureRefine, PartDesign::Feature)
+
+FeatureRefine::FeatureRefine()
+{
+    ADD_PROPERTY_TYPE(Refine,
+                      (0),
+                      "Part Design",
+                      (App::PropertyType)(App::Prop_None),
+                      "Refine shape (clean up redundant edges) after operations");
+    // init Refine property
+    Base::Reference<ParameterGrp> hGrp = App::GetApplication()
+                                             .GetUserParameter()
+                                             .GetGroup("BaseApp")
+                                             ->GetGroup("Preferences")
+                                             ->GetGroup("Mod/PartDesign");
+    this->Refine.setValue(hGrp->GetBool("RefineModel", true));
+}
+
+bool FeatureRefine::onlyHasToRefine() const
+{
+    if (!Refine.isTouched()) {
+        return false;
+    }
+    if (rawShape.isNull()) {
+        return false;
+    }
+    std::vector<App::Property*> propList;
+    getPropertyList(propList);
+    for (auto prop : propList) {
+        if (prop != &Refine
+            /*&& prop != &SuppressedShape*/
+            && prop->isTouched()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+bool FeatureRefine::onlyHaveRefined()
+{
+    if (onlyHasToRefine()) {
+        TopoShape result = refineShapeIfActive(rawShape);
+        Shape.setValue(result);
+        return true;
+    }
+    return false;
+}
+
```

Comment:
```suggestion
bool FeatureRefine::onlyHaveRefined()
{
    if (!Refine.isTouched()) {
        return false;
    }
    if (rawShape.isNull()) {
        return false;
    }
    std::vector<App::Property*> propList;
    getPropertyList(propList);
    for (auto prop : propList) {
        if (prop != &Refine
            /*&& prop != &SuppressedShape*/
            && prop->isTouched()) {
            return false;
        }
    }
    TopoShape result = refineShapeIfActive(rawShape);
    Shape.setValue(result);
    return true;
}
```

`onlyHasToRefine` is not used elsewhere so we can concatenate.